### PR TITLE
Update the operator-sdk 

### PR DESCRIPTION
**Pull request by ChunxiAlexLuo**
Signed-off-by: Chunxi Luo chuluo@redhat.com

https://github.com/open-cluster-management/backlog/issues/15549

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -214,6 +214,7 @@ ifndef DOCKER_PASS
 endif
 
 kind-deploy-controller: check-env
+	kubectl create ns $(KIND_NAMESPACE) --kubeconfig=$(PWD)/kubeconfig_hub
```
Comment (ChunxiAlexLuo) [inline Makefile:217]: It seems we also need to create `open-cluster-management-agent-addon` ns on the hub cluster for leader election, correct me if I'm wrong

**Comment by ChunxiAlexLuo** [comment]
@mprahl @JustinKuli  here the first pr is what I did before our yesterday discussion, I don't squash the second pr `fix e2e testing` for easier reviewing

File: main.go [hunk 2]
```diff
@@ -13,23 +12,29 @@ import (
...
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
-	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
```
Comment (mprahl) [inline main.go:39]: Could you please add the following comment after this import?
```go
//+kubebuilder:scaffold:imports
```

File: test/e2e/case1_spec_sync_test.go [hunk 3]
```diff
@@ -18,14 +18,15 @@ const case1PolicyYaml string = "../resources/case1_spec_sync/case1-test-policy.y
 var _ = Describe("Test spec sync", func() {
 	BeforeEach(func() {
 		By("Creating a policy on hub cluster in ns:" + testNamespace)
-		utils.Kubectl("apply", "-f", case1PolicyYaml, "-n", testNamespace,
+		_, err := utils.KubectlWithOutput("apply", "-f", case1PolicyYaml, "-n", testNamespace,
```
Comment (mprahl) [inline test/e2e/case1_spec_sync_test.go:21]: Could you please explain why this change was made in the commit message? Ideally this change would have been in a separate commit but an explanation would suffice. I'm pretty sure I know why you did it but someone without context would not and so looking at the commit message would help them.

File: Makefile [hunk 4]
```diff
@@ -158,6 +158,43 @@ clean::
...
+CONTROLLER_GEN = $(shell pwd)/bin/controller-gen
+KUSTOMIZE = $(shell pwd)/bin/kustomize
+
+.PHONY: manifests
+manifests: controller-gen
+	$(CONTROLLER_GEN) $(CRD_OPTIONS) rbac:roleName=governance-policy-spec-sync paths="./..." output:rbac:artifacts:config=deploy/rbac
+
+.PHONY: generate
+generate: controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
+	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
```
Comment (mprahl) [inline Makefile:173]: `hack/boilerplate.go.txt` seems to be missing. Could you please add it?

File: test/resources/case2_uninstall_ns/case2-uninstall-ns.yaml [hunk 5]
```diff
@@ -27,7 +27,7 @@ spec:
       serviceAccountName: governance-policy-spec-sync
       containers:
       - name: uninstall-ns
-        image: "quay.io/open-cluster-management/governance-policy-spec-sync:latest"
+        image: quay.io/open-cluster-management/governance-policy-spec-sync:latest
```
Comment (mprahl) [inline test/resources/case2_uninstall_ns/case2-uninstall-ns.yaml:30]: In the future, it'd be nice to not include things that aren't related to your PR and that don't affect the behavior of the code at all.

File: main.go [hunk 6]
```diff
@@ -123,29 +142,26 @@ func main() {
...
-		log.Error(err, "")
+	if err = (&sync.PolicyReconciler{
+		HubClient:       mgr.GetClient(),
+		ManagedClient:   managedClient,
+		ManagedRecorder: managedRecorder,
+		Scheme:          mgr.GetScheme(),
+	}).SetupWithManager(mgr); err != nil {
+		log.Error(err, "unable to create controller", "controller", sync.ControllerName)
 		os.Exit(1)
 	}
```
Comment (mprahl) [inline main.go:162]: Could you please add the health and readiness checks? For example like this?
https://github.com/open-cluster-management/governance-policy-status-sync/blob/81abad3c816d085c8755b4e6122a6336dfa0a743/main.go#L175-L183

Then you also need to modify the `Deployment` in the `deploy` directory to use them.

File: main.go [hunk 7]
```diff
@@ -83,34 +112,24 @@ func main() {
...
-		log.Error(err, "")
-		os.Exit(1)
-	}
+	var kubeClient kubernetes.Interface = kubernetes.NewForConfigOrDie(managedCfg)
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events(namespace)})
+	managedRecorder := eventBroadcaster.NewRecorder(eventsScheme, v1.EventSource{Component: sync.ControllerName})
 
 	// Set default manager options
 	options := manager.Options{
```
Comment (mprahl) [inline main.go:127]: Could you also please add the ability to fallback to the legacy leader election as shown in this PR?
https://github.com/open-cluster-management/cert-policy-controller/pull/105

This is required for this controller to run on OCP 3.11.

**Review by mprahl — CHANGES_REQUESTED** [review]

**Comment by ChunxiAlexLuo** [comment]
health and readiness checks will be added another pr

https://github.com/open-cluster-management/governance-policy-spec-sync/pull/65

File: main.go [hunk 8]
```diff
@@ -123,29 +153,26 @@ func main() {
...
-		log.Error(err, "")
+	if err = (&sync.PolicyReconciler{
+		HubClient:       mgr.GetClient(),
+		ManagedClient:   managedClient,
+		ManagedRecorder: managedRecorder,
+		Scheme:          mgr.GetScheme(),
+	}).SetupWithManager(mgr); err != nil {
+		log.Error(err, "unable to create controller", "controller", sync.ControllerName)
 		os.Exit(1)
 	}
```
Comment (mprahl) [inline main.go:173]: After this line, could you please include the `//+kubebuilder:scaffold:builder` comment?

File: controllers/sync/policy_spec_sync.go [hunk 9]
```diff
@@ -7,110 +7,59 @@ import (
...
+	Scheme          *runtime.Scheme
 }
 
 // Reconcile reads that state of the cluster for a Policy object and makes changes based on the state read
 // and what is in the Policy.Spec
 // Note:
 // The Controller will requeue the Request to be processed again if the returned error is non-nil or
 // Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
-func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+func (r *PolicyReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
```
Comment (mprahl) [inline controllers/sync/policy_spec_sync.go:51]: I don't see any kubebuilder RBAC markers in this file (i.e. `//+kubebuilder:rbac`). Could you please add them to match what was originally in `deploy/operator.yaml`? It seems the `deploy/rbac/role.yaml` doesn't match what was there previously.

File: main.go [hunk 10]
```diff
@@ -83,34 +114,33 @@ func main() {
...
 
 	// Set default manager options
 	options := manager.Options{
+		Scheme:             scheme,
 		Namespace:          namespace,
 		MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
+		LeaderElection:     tool.Options.EnableLeaderElection,
+		LeaderElectionID:   "policy-spec-sync.open-cluster-management.io",
+		// Override LeaderElectionConfig to managed cluster config.
+		// Otherwise it will improperly use hub cluster config as default on ceating a new manager step.
```
Comment (mprahl) [inline main.go:136]:
```suggestion
		// Otherwise it will improperly use the hub cluster config for leader election.
```

**Review by mprahl — CHANGES_REQUESTED** [review]
Could you also please modify your commit message subject line to something like `Update the operator-sdk` instead of `update-sdk-for-governance-policy-spec-sync`?

**Comment by ChunxiAlexLuo** [comment]
@mprahl 
Don't get what should be matched with `deploy/operator.yaml `in `ClusterRole`, current `role.yaml `looks fine to me, you mean should use `asterisk` for referring all resources under `policy.open-cluster-management.io`? Updated the rest as the suggestions

**Comment by mprahl** [comment]
> @mprahl Don't get what should be matched with `deploy/operator.yaml `in `ClusterRole`, current `role.yaml `looks fine to me, you mean should use `asterisk` for referring all resources under `policy.open-cluster-management.io`? Updated the rest as the suggestions

I must have misread it. With your latest change with the RBAC kubebuilder markers and the CI verifying, this should be fine!

**Review by mprahl — APPROVED** [review]
Nice job! This looks good to me.

/hold in case you want someone else to review it. If not, feel free to unhold it.

**Pull request merged**